### ISSUE TITLE
Support for oneOf keyword

### ISSUE THREAD

**Issue by Nickcw6** [pactflow/swagger-mock-validator#21]
Following up from this thread: [https://pact-foundation.slack.com/archives/CLS16AVEE/p1681405717867409](https://pact-foundation.slack.com/archives/CLS16AVEE/p1681405717867409)

Would it be possible to support `oneOf`/`anyOf` keywords in OAS schemas please? Or failing that, any guidance on workarounds? The current [documentation](https://docs.pactflow.io/docs/bi-directional-contract-testing/contracts/oas/keyword-support/) implies this should be possible in a similar manner to `allOf`, but unfortunately this still results in a bunch of `should NOT have additional properties` errors. 

There does seem to be a workaround by adding `additionalProperties: true` wherever there a `oneOf` keyword is used, but this in doing so you're losing some of the benefits contract testing provides. 

Have attached an example OAS file, a merged & derefernced version (using [this](https://github.com/YOU54F/oas-merge) tool) & Pact file: [oas-pact-oneof-example.zip](https://github.com/pactflow/swagger-mock-validator/files/11357417/oas-pact-oneof-example.zip)

Command ran for comparison:

`swagger-mock-validator oas/vets_example_oas-transformed.json pacts/vetbookings-vetclinic.json`

Errors received: 

```
Mock file "pacts/vetbookings-vetclinic.json" is not compatible with spec file "oas/vets_example_oas-transformed.json"
7 error(s)
        response.body.incompatible: 7
0 warning(s)
{
  warnings: [],
  errors: [
    {
      code: 'response.body.incompatible',
      message: 'Response body is incompatible with the response body schema in the spec file: should NOT have additional properties - id',
      mockDetails: {
        interactionDescription: 'A request to view owner 9b4f0df1-e41b-497d-b740-e1c65b89734c',
        interactionState: 'There is a pet owner with ID 9b4f0df1-e41b-497d-b740-e1c65b89734c',
        location: '[root].interactions[0].response.body.pets[0]',
        mockFile: 'pacts/vetbookings-vetclinic.json',
        value: {
          id: '6366c28a-1a6e-4233-a56a-bb902060d642',
          name: 'Dave',
          age: 4,
          species: 'Cat',
          discharged: false,
          colour: 'Black',
          mischievous_scale: 10
        }
      },
      source: 'spec-mock-validation',
      specDetails: {
        location: '[root].paths./clinic/owner/{id}.get.responses.200.content.application/json.schema.properties.pets.items.additionalProperties',
        pathMethod: 'get',
        pathName: '/clinic/owner/{id}',
        specFile: 'oas/vets_example_oas-transformed.json',
        value: undefined
      },
      type: 'error'
    },
    {
      code: 'response.body.incompatible',
      message: 'Response body is incompatible with the response body schema in the spec file: should NOT have additional properties - name',
      mockDetails: {
        interactionDescription: 'A request to view owner 9b4f0df1-e41b-497d-b740-e1c65b89734c',
        interactionState: 'There is a pet owner with ID 9b4f0df1-e41b-497d-b740-e1c65b89734c',
        location: '[root].interactions[0].response.body.pets[0]',
        mockFile: 'pacts/vetbookings-vetclinic.json',
        value: {
          id: '6366c28a-1a6e-4233-a56a-bb902060d642',
          name: 'Dave',
          age: 4,
          species: 'Cat',
          discharged: false,
          colour: 'Black',
          mischievous_scale: 10
        }
      },
      source: 'spec-mock-validation',
      specDetails: {
        location: '[root].paths./clinic/owner/{id}.get.responses.200.content.application/json.schema.properties.pets.items.additionalProperties',
        pathMethod: 'get',
        pathName: '/clinic/owner/{id}',
        specFile: 'oas/vets_example_oas-transformed.json',
        value: undefined
      },
      type: 'error'
    },
    {
      code: 'response.body.incompatible',
      message: 'Response body is incompatible with the response body schema in the spec file: should NOT have additional properties - age',
      mockDetails: {
        interactionDescription: 'A request to view owner 9b4f0df1-e41b-497d-b740-e1c65b89734c',
        interactionState: 'There is a pet owner with ID 9b4f0df1-e41b-497d-b740-e1c65b89734c',
        location: '[root].interactions[0].response.body.pets[0]',
        mockFile: 'pacts/vetbookings-vetclinic.json',
        value: {
          id: '6366c28a-1a6e-4233-a56a-bb902060d642',
          name: 'Dave',
          age: 4,
          species: 'Cat',
          discharged: false,
          colour: 'Black',
          mischievous_scale: 10
        }
      },
      source: 'spec-mock-validation',
      specDetails: {
        location: '[root].paths./clinic/owner/{id}.get.responses.200.content.application/json.schema.properties.pets.items.additionalProperties',
        pathMethod: 'get',
        pathName: '/clinic/owner/{id}',
        specFile: 'oas/vets_example_oas-transformed.json',
        value: undefined
      },
      type: 'error'
    },
    {
      code: 'response.body.incompatible',
      message: 'Response body is incompatible with the response body schema in the spec file: should NOT have additional properties - species',
      mockDetails: {
        interactionDescription: 'A request to view owner 9b4f0df1-e41b-497d-b740-e1c65b89734c',
        interactionState: 'There is a pet owner with ID 9b4f0df1-e41b-497d-b740-e1c65b89734c',
        location: '[root].interactions[0].response.body.pets[0]',
        mockFile: 'pacts/vetbookings-vetclinic.json',
        value: {
          id: '6366c28a-1a6e-4233-a56a-bb902060d642',
          name: 'Dave',
          age: 4,
          species: 'Cat',
          discharged: false,
          colour: 'Black',
          mischievous_scale: 10
        }
      },
      source: 'spec-mock-validation',
      specDetails: {
        location: '[root].paths./clinic/owner/{id}.get.responses.200.content.application/json.schema.properties.pets.items.additionalProperties',
        pathMethod: 'get',
        pathName: '/clinic/owner/{id}',
        specFile: 'oas/vets_example_oas-transformed.json',
        value: undefined
      },
      type: 'error'
    },
    {
      code: 'response.body.incompatible',
      message: 'Response body is incompatible with the response body schema in the spec file: should NOT have additional properties - discharged',
      mockDetails: {
        interactionDescription: 'A request to view owner 9b4f0df1-e41b-497d-b740-e1c65b89734c',
        interactionState: 'There is a pet owner with ID 9b4f0df1-e41b-497d-b740-e1c65b89734c',
        location: '[root].interactions[0].response.body.pets[0]',
        mockFile: 'pacts/vetbookings-vetclinic.json',
        value: {
          id: '6366c28a-1a6e-4233-a56a-bb902060d642',
          name: 'Dave',
          age: 4,
          species: 'Cat',
          discharged: false,
          colour: 'Black',
          mischievous_scale: 10
        }
      },
      source: 'spec-mock-validation',
      specDetails: {
        location: '[root].paths./clinic/owner/{id}.get.responses.200.content.application/json.schema.properties.pets.items.additionalProperties',
        pathMethod: 'get',
        pathName: '/clinic/owner/{id}',
        specFile: 'oas/vets_example_oas-transformed.json',
        value: undefined
      },
      type: 'error'
    },
    {
      code: 'response.body.incompatible',
      message: 'Response body is incompatible with the response body schema in the spec file: should NOT have additional properties - colour',
      mockDetails: {
        interactionDescription: 'A request to view owner 9b4f0df1-e41b-497d-b740-e1c65b89734c',
        interactionState: 'There is a pet owner with ID 9b4f0df1-e41b-497d-b740-e1c65b89734c',
        location: '[root].interactions[0].response.body.pets[0]',
        mockFile: 'pacts/vetbookings-vetclinic.json',
        value: {
          id: '6366c28a-1a6e-4233-a56a-bb902060d642',
          name: 'Dave',
          age: 4,
          species: 'Cat',
          discharged: false,
          colour: 'Black',
          mischievous_scale: 10
        }
      },
      source: 'spec-mock-validation',
      specDetails: {
        location: '[root].paths./clinic/owner/{id}.get.responses.200.content.application/json.schema.properties.pets.items.additionalProperties',
        pathMethod: 'get',
        pathName: '/clinic/owner/{id}',
        specFile: 'oas/vets_example_oas-transformed.json',
        value: undefined
      },
      type: 'error'
    },
    {
      code: 'response.body.incompatible',
      message: 'Response body is incompatible with the response body schema in the spec file: should NOT have additional properties - mischievous_scale',
      mockDetails: {
        interactionDescription: 'A request to view owner 9b4f0df1-e41b-497d-b740-e1c65b89734c',
        interactionState: 'There is a pet owner with ID 9b4f0df1-e41b-497d-b740-e1c65b89734c',
        location: '[root].interactions[0].response.body.pets[0]',
        mockFile: 'pacts/vetbookings-vetclinic.json',
        value: {
          id: '6366c28a-1a6e-4233-a56a-bb902060d642',
          name: 'Dave',
          age: 4,
          species: 'Cat',
          discharged: false,
          colour: 'Black',
          mischievous_scale: 10
        }
      },
      source: 'spec-mock-validation',
      specDetails: {
        location: '[root].paths./clinic/owner/{id}.get.responses.200.content.application/json.schema.properties.pets.items.additionalProperties',
        pathMethod: 'get',
        pathName: '/clinic/owner/{id}',
        specFile: 'oas/vets_example_oas-transformed.json',
        value: undefined
      },
      type: 'error'
    }
  ]
}

Error: Mock file "pacts/vetbookings-vetclinic.json" is not compatible with spec file "oas/vets_example_oas-transformed.json"
```

Thanks!

**Comment by mefellows** [comment]
Hi Nick, we need to review the details of this to understand why it's not working as expected. We should be able to support this, however the way JSON schemas are applied in certain scenarios makes the application of the input to the schemas mutually exclusive, resulting in the errors you see.

I don't have a timeline to share with you just yet, but we will be reviewing it in this quarter.

**Comment by mefellows** [comment]
@Nickcw6 in investigating a similar issue with another contract, I stumbled upon the same conclusion you came to (I clearly missed the insight you added, but probably best I did!)

> There does seem to be a workaround by adding additionalProperties: true wherever there a oneOf keyword is used, but this in doing so you're losing some of the benefits contract testing provides.

I'm not sure what benefits are actually lost in this case. I suspect there is some bug in the validator because when you add it, it seems to do what we want:

OAS:
```yml
openapi: 3.0.1
info:
  title: anyOf / oneOf example
  description: Demonstrate the anyOf and oneOf issue
  version: 1.0.0
paths:
  /products/{id}:
    get:
      summary: Find product by ID
      description: Returns a single product
      operationId: getProductByID
      parameters:
      - name: id
        in: path
        description: ID of product to get
        schema:
          type: string
        required: true
        example: "10"
      responses:
        "200":
          description: successful operation
          content:
            "application/json":
              schema:
                type: "object"
                additionalProperties: true
                oneOf:
                  - $ref: '#/components/schemas/Cat'
                  - $ref: '#/components/schemas/Product'
              examples:
                application/json:
                  value:
                    id: "1234"
                    type: "food"
                    price: 42
                    name: "thing"
        "400":
          description: Invalid ID supplied
          content: {}
        "404":
          description: Product not found
          content: {}
components:
  schemas:
    Product:
      type: object
      required:
        - name
        - price
        - id
        - type
        - version
      additionalProperties: false
      properties:
        id:
          type: string
        type:
          type: string
        name:
          type: string
        version:
          type: string
        price:
          type: number
    Cat:
      type: object
      required:
        - name
        - owner
      additionalProperties: false
      properties:
        name:
          type: string
        owner:
          type: string
```

Pact file:

```json
{
  "consumer": {
    "name": "pactflow-example-consumer"
  },
  "provider": {
    "name": "collaborative-contracts-provider"
  },
  "interactions": [
    {
      "description": "a request to get a product",
      "request": {
        "method": "GET",
        "path": "/products/10",
        "headers": {
          "Authorization": "Bearer 2019-01-14T11:34:18.045Z"
        }
      },
      "response": {
        "status": 200,
        "headers": {
          "Content-Type": "application/json"
        },
        "body": {
          "name": "10",
          "type": "CREDIT_CARD",
          "price": 28
        }
      }
    },
    {
      "description": "a request to get a cat shaped product",
      "request": {
        "method": "GET",
        "path": "/products/cat",
        "headers": {
          "Authorization": "Bearer 2019-01-14T11:34:18.045Z"
        }
      },
      "response": {
        "status": 200,
        "headers": {
          "Content-Type": "application/json"
        },
        "body": {
          "name": "kitty",
          "owner": "crazy cat lady"
        }
      }
    }
  ]
}
```

If you were to modify one of the interactions to have a key that doesn't exist, the test fails as expect.d

**Comment by mefellows** [comment]
Thanks Nick - great to hear that helped. We've continuing to make changes to this over the next little while so I'm compiling a list of examples to elucidate the behaviour.

I'll link that from the README here, and also over in the PactFlow docs in the coming weeks.

Any additional feedback on the above please do feel free to update us here.

**Comment by mefellows** [comment]
I believe this was fixed (and improved) in #32. 

The examples in that linked repo have been updated accordingly also.

Closing this one now.
